Add an optional ACME certificate provider.

jmalloc opened the pull request:
This PR adds support for ACME-based certificate generation.

This required a fairly severe overhaul of the certificate "provider" system, which included consolidating the per-provider caches into a single global certificate cache.

There are now two "chains" of certificate providers, one for server names that are "recognized", that is, they are mapped to some known Docker service, and another for those that are "unrecognized".

When a certificate is obtained, it is given a rank based on the provider that was used to obtain it. The providers for "unrecognized" server names are always ranked worse (numerically higher).

When a certificate is requested from the cache, a "worst acceptable rank" value is provided. If the cache certificate has a worse rank than this value, the cache entry is ignored.

The upshot of all of this is that:

- certificates from on-disk files, or from ACME take precedence over generated certs from the Honeycomb CA
- certificates that are obtained for an "unrecognized" server name are discarded if that server name later becomes "recognized" - This can happen if you hit a newly deployed service's public domain before Honeycomb detects the service's existence

By default, the provider "chains" for both recognized and unrecognized server names consists of the `FileProvider` and the `AdhocProvider`.

If the environment variable `ACME_EMAIL` is defined, the `ACMEProvider` is inserted between these two existing providers for the recognized chain.

The `ACME_CACHE` environment variable is the path to directory used to store the certificates obtained via ACME.  According to the `autocert` docs, this can be shared by multiple managers, which I hope/take to mean that multiple Honeycomb instances can share the cache via Docker volumes.

Finally the `ACME_URL` environment variable sets the endpoint for the ACME API.  If this value is undefined it defaults to the Let's Encrypt production API.

jmalloc committed:
Add an optional ACME certificate provider.

jmalloc committed:
Add `ACME_URL` variable.

jmalloc committed:
Reimplement certificate provider system to use a single ranked cache.

jmalloc commented:
As far as I know, right at this moment, the code is "complete", but I will need to do further testing in the company sandbox next week.

ezzatron commented on src/cmd/honeycomb/certs.go:
```diff
@@ -0,0 +1,182 @@
+	}, true
+}
+
+// fileCertificateProvider returns a cert provider that obtains certificates
```
Should be `acmeCertificateProvider`?

ezzatron reviewed (commented):
I only gave it a quick scan, but seems logical 👍

jmalloc committed:
Use correct function name in doc block.

jmalloc committed:
Remove unused `Domains` field from ACME configuration.

koshatul commented:
I think I've said it before, I'm a huge fan of the `.next()` style of ordering processors instead of using a slice of them.

Other than that I'd need to sit down and read through a lot of changes :)

jmalloc commented:
Yeah, it's pretty involved.  And I can say right now, that it doesn't work :P

jmalloc commented:
> I think I've said it before, I'm a huge fan of the `.next()` style of ordering processors instead of using a slice of them.

I use that approach when implementing middlewares - things that can conditionally forward onto the "next stage".

In this case, it's a priority ordered list, it doesn't make sense for each implementation to know about the others. They can also appear in multiple lists with different priorities, meaning conceptually there are different "next" providers to try depending on other factors (whether or not the domain is recognised).

jmalloc committed:
Added TTL to log output for acquired certificates.

jmalloc committed:
Panic if an uncachable certificate provider result is added to the cache.

jmalloc committed:
Fix `AdhocProvider.IsValid()` expirty checking logic (date comparison was backwards).

jmalloc committed:
Only remove certificate cache entries that are invalidated by their provider.

Prior to this change, a cache-search that failed because the rank requirement was not met
also caused an invalidation.

jmalloc committed:
Add logging when cache entries are added and removed.

jmalloc committed:
Move common certificate log formating into utility function.

jmalloc committed:
Implement `cert.Resolver.IsRecognized` hook.

koshatul commented on src/frontend/cert/file_provider.go:
```diff
@@ -64,12 +63,9 @@ func (p *FileProvider) GetCertificate(
 			p.Logger.Printf(
-				"Loaded certificate for '%s' from '%s', expires at %s (%s), issued by '%s'",
-				n.Unicode,
+				"Loaded certificate %s from from '%s'",
```
`from from`

koshatul reviewed (changes requested):
just the `from from`, everything else looks OK, I might need to check out the branch to actually follow the new code though.
